In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
df = pd.read_csv(r'C:\Users\Avery\OneDrive\Documents\GitHub\Group10_Final_Project\201306-citibike-tripdata.csv')

df.head(5)

tripduration            starttime             stoptime  start station id  \
0           695  2013-06-01 00:00:01  2013-06-01 00:11:36               444   
1           693  2013-06-01 00:00:08  2013-06-01 00:11:41               444   
2          2059  2013-06-01 00:00:44  2013-06-01 00:35:03               406   
3           123  2013-06-01 00:01:04  2013-06-01 00:03:07               475   
4          1521  2013-06-01 00:01:22  2013-06-01 00:26:43              2008   

       start station name  start station latitude  start station longitude  \
0      Broadway & W 24 St               40.742354               -73.989151   
1      Broadway & W 24 St               40.742354               -73.989151   
2  Hicks St & Montague St               40.695128               -73.995951   
3     E 15 St & Irving Pl               40.735243               -73.987586   
4   Little West St & 1 Pl               40.705693               -74.016777   

   end station id        end station name  end station latitude  \
0           434.0         9 Ave & W 18 St             40.743174   
1           434.0         9 Ave & W 18 St             40.743174   
2           406.0  Hicks St & Montague St             40.695128   
3           262.0         Washington Park             40.691782   
4           310.0     State St & Smith St             40.689269   

   end station longitude  bikeid    usertype  birth year  gender  
0             -74.003664   19678  Subscriber      1983.0       1  
1             -74.003664   16649  Subscriber      1984.0       1  
2             -73.995951   19599    Customer         NaN       0  
3             -73.973730   16352  Subscriber      1960.0       1  
4             -73.989129   15567  Subscriber      1983.0       1

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 577703 entries, 0 to 577702
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   tripduration             577703 non-null  int64  
 1   starttime                577703 non-null  object 
 2   stoptime                 577703 non-null  object 
 3   start station id         577703 non-null  int64  
 4   start station name       577703 non-null  object 
 5   start station latitude   577703 non-null  float64
 6   start station longitude  577703 non-null  float64
 7   end station id           559644 non-null  float64
 8   end station name         559644 non-null  object 
 9   end station latitude     559644 non-null  float64
 10  end station longitude    559644 non-null  float64
 11  bikeid                   577703 non-null  int64  
 12  usertype                 577703 non-null  object 
 13  birth year               337382 non-null  float64
 14  gend

Now using the df.info() we are able to see somethings that must be done before visualizing and implementing machine learning.
1) We can shorten the names of the columns for easier access.
2) Find any null entries that are in important areas such as the time or stations.

In [4]:
df = df.rename(columns={'start station id':'ss_id', 'start station name':'ss_name', 'start station latitude':'ss_latitude', 'start station longitude':'ss_longitude',
                        'end station id':'es_id', 'end station name':'es_name', 'end station latitude':'es_latitude', 'end station longitude':'es_longitude'})

df.head(5)

tripduration            starttime             stoptime  ss_id  \
0           695  2013-06-01 00:00:01  2013-06-01 00:11:36    444   
1           693  2013-06-01 00:00:08  2013-06-01 00:11:41    444   
2          2059  2013-06-01 00:00:44  2013-06-01 00:35:03    406   
3           123  2013-06-01 00:01:04  2013-06-01 00:03:07    475   
4          1521  2013-06-01 00:01:22  2013-06-01 00:26:43   2008   

                  ss_name  ss_latitude  ss_longitude  es_id  \
0      Broadway & W 24 St    40.742354    -73.989151  434.0   
1      Broadway & W 24 St    40.742354    -73.989151  434.0   
2  Hicks St & Montague St    40.695128    -73.995951  406.0   
3     E 15 St & Irving Pl    40.735243    -73.987586  262.0   
4   Little West St & 1 Pl    40.705693    -74.016777  310.0   

                  es_name  es_latitude  es_longitude  bikeid    usertype  \
0         9 Ave & W 18 St    40.743174    -74.003664   19678  Subscriber   
1         9 Ave & W 18 St    40.743174    -74.003664   16649  Subscriber   
2  Hicks St & Montague St    40.695128    -73.995951   19599    Customer   
3         Washington Park    40.691782    -73.973730   16352  Subscriber   
4     State St & Smith St    40.689269    -73.989129   15567  Subscriber   

   birth year  gender  
0      1983.0       1  
1      1984.0       1  
2         NaN       0  
3      1960.0       1  
4      1983.0       1

In [5]:
df.isnull().sum()

tripduration         0
starttime            0
stoptime             0
ss_id                0
ss_name              0
ss_latitude          0
ss_longitude         0
es_id            18059
es_name          18059
es_latitude      18059
es_longitude     18059
bikeid               0
usertype             0
birth year      240321
gender               0
dtype: int64

Using the above table there are instances that need to be noticed:
1) There are 240321 instances where the birth year is not record. As we are likely not looking into this fact, it does not need to be changed. 
2) There are 18059 instances where there is no end station. Atleast, the number of number of instances accross all end stations are all the same value. As these areas need to be filled, we will create our own value to be filled that will be common throughout our datframe. This is essential to call back these insatnces later for further analysis.

In [6]:
unique_station_ids = df['es_id'].unique()

if '0.0' in unique_station_ids:
    print('0.0 exists')

C:\Users\Avery\AppData\Local\Temp\ipykernel_18160\3709806180.py:3: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  if '0.0' in unique_station_ids:


Since 0.0 does not exist in the end station category thus we can name the station id 0.0
We can name the latitude and longitude bot as 0.0 as well as New York (where the data was taken) is now where near these coordinates. We will name the null values in the end station name column as 'no_return'. This will allow us to ensure there is no compromise to the data and that there is no null rows. This esures that further analysis can be done by the company into why there is no end station data for these instances and will not interfer with the machine learning later on. 

In [7]:
df['es_name'] = df['es_name'].fillna('no_return')

df.isnull().sum()

tripduration         0
starttime            0
stoptime             0
ss_id                0
ss_name              0
ss_latitude          0
ss_longitude         0
es_id            18059
es_name              0
es_latitude      18059
es_longitude     18059
bikeid               0
usertype             0
birth year      240321
gender               0
dtype: int64

In [8]:
df = df.fillna(0)

df.isnull().sum()

tripduration    0
starttime       0
stoptime        0
ss_id           0
ss_name         0
ss_latitude     0
ss_longitude    0
es_id           0
es_name         0
es_latitude     0
es_longitude    0
bikeid          0
usertype        0
birth year      0
gender          0
dtype: int64

Now we can change the types to better reflect the values stored.

In [9]:
df = df.astype({"es_id": int, 'birth year': int})

It is also prudent to seperate the start and end times seperate from the dates, and to seperate the date formats into a seperate column for month, day and year

In [10]:
df[['startdate', 'starttime']] = df['starttime'].str.split(' ', expand=True)

df.head()

tripduration starttime             stoptime  ss_id                 ss_name  \
0           695  00:00:01  2013-06-01 00:11:36    444      Broadway & W 24 St   
1           693  00:00:08  2013-06-01 00:11:41    444      Broadway & W 24 St   
2          2059  00:00:44  2013-06-01 00:35:03    406  Hicks St & Montague St   
3           123  00:01:04  2013-06-01 00:03:07    475     E 15 St & Irving Pl   
4          1521  00:01:22  2013-06-01 00:26:43   2008   Little West St & 1 Pl   

   ss_latitude  ss_longitude  es_id                 es_name  es_latitude  \
0    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
1    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
2    40.695128    -73.995951    406  Hicks St & Montague St    40.695128   
3    40.735243    -73.987586    262         Washington Park    40.691782   
4    40.705693    -74.016777    310     State St & Smith St    40.689269   

   es_longitude  bikeid    usertype  birth year  gender   startdate  
0    -74.003664   19678  Subscriber        1983       1  2013-06-01  
1    -74.003664   16649  Subscriber        1984       1  2013-06-01  
2    -73.995951   19599    Customer           0       0  2013-06-01  
3    -73.973730   16352  Subscriber        1960       1  2013-06-01  
4    -73.989129   15567  Subscriber        1983       1  2013-06-01

In [11]:
df['startdate'] = pd.to_datetime(df['startdate'])

df['day_of_week'] = df['startdate'].dt.day_name()

df.head(5)

tripduration starttime             stoptime  ss_id                 ss_name  \
0           695  00:00:01  2013-06-01 00:11:36    444      Broadway & W 24 St   
1           693  00:00:08  2013-06-01 00:11:41    444      Broadway & W 24 St   
2          2059  00:00:44  2013-06-01 00:35:03    406  Hicks St & Montague St   
3           123  00:01:04  2013-06-01 00:03:07    475     E 15 St & Irving Pl   
4          1521  00:01:22  2013-06-01 00:26:43   2008   Little West St & 1 Pl   

   ss_latitude  ss_longitude  es_id                 es_name  es_latitude  \
0    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
1    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
2    40.695128    -73.995951    406  Hicks St & Montague St    40.695128   
3    40.735243    -73.987586    262         Washington Park    40.691782   
4    40.705693    -74.016777    310     State St & Smith St    40.689269   

   es_longitude  bikeid    usertype  birth year  gender  startdate day_of_week  
0    -74.003664   19678  Subscriber        1983       1 2013-06-01    Saturday  
1    -74.003664   16649  Subscriber        1984       1 2013-06-01    Saturday  
2    -73.995951   19599    Customer           0       0 2013-06-01    Saturday  
3    -73.973730   16352  Subscriber        1960       1 2013-06-01    Saturday  
4    -73.989129   15567  Subscriber        1983       1 2013-06-01    Saturday

In [12]:
df['startdate'] = df['startdate'].astype(str)

df.head(5)

tripduration starttime             stoptime  ss_id                 ss_name  \
0           695  00:00:01  2013-06-01 00:11:36    444      Broadway & W 24 St   
1           693  00:00:08  2013-06-01 00:11:41    444      Broadway & W 24 St   
2          2059  00:00:44  2013-06-01 00:35:03    406  Hicks St & Montague St   
3           123  00:01:04  2013-06-01 00:03:07    475     E 15 St & Irving Pl   
4          1521  00:01:22  2013-06-01 00:26:43   2008   Little West St & 1 Pl   

   ss_latitude  ss_longitude  es_id                 es_name  es_latitude  \
0    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
1    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
2    40.695128    -73.995951    406  Hicks St & Montague St    40.695128   
3    40.735243    -73.987586    262         Washington Park    40.691782   
4    40.705693    -74.016777    310     State St & Smith St    40.689269   

   es_longitude  bikeid    usertype  birth year  gender   startdate  \
0    -74.003664   19678  Subscriber        1983       1  2013-06-01   
1    -74.003664   16649  Subscriber        1984       1  2013-06-01   
2    -73.995951   19599    Customer           0       0  2013-06-01   
3    -73.973730   16352  Subscriber        1960       1  2013-06-01   
4    -73.989129   15567  Subscriber        1983       1  2013-06-01   

  day_of_week  
0    Saturday  
1    Saturday  
2    Saturday  
3    Saturday  
4    Saturday

In [13]:


df[['startyear', 'startmonth', 'startdate']] = df['startdate'].str.split('-', expand=True)

df.head()

tripduration starttime             stoptime  ss_id                 ss_name  \
0           695  00:00:01  2013-06-01 00:11:36    444      Broadway & W 24 St   
1           693  00:00:08  2013-06-01 00:11:41    444      Broadway & W 24 St   
2          2059  00:00:44  2013-06-01 00:35:03    406  Hicks St & Montague St   
3           123  00:01:04  2013-06-01 00:03:07    475     E 15 St & Irving Pl   
4          1521  00:01:22  2013-06-01 00:26:43   2008   Little West St & 1 Pl   

   ss_latitude  ss_longitude  es_id                 es_name  es_latitude  \
0    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
1    40.742354    -73.989151    434         9 Ave & W 18 St    40.743174   
2    40.695128    -73.995951    406  Hicks St & Montague St    40.695128   
3    40.735243    -73.987586    262         Washington Park    40.691782   
4    40.705693    -74.016777    310     State St & Smith St    40.689269   

   es_longitude  bikeid    usertype  birth year  gender startdate day_of_week  \
0    -74.003664   19678  Subscriber        1983       1        01    Saturday   
1    -74.003664   16649  Subscriber        1984       1        01    Saturday   
2    -73.995951   19599    Customer           0       0        01    Saturday   
3    -73.973730   16352  Subscriber        1960       1        01    Saturday   
4    -73.989129   15567  Subscriber        1983       1        01    Saturday   

  startyear startmonth  
0      2013         06  
1      2013         06  
2      2013         06  
3      2013         06  
4      2013         06

In [14]:
df = df[['tripduration', 'startyear', 'startmonth', 'startdate', 'starttime','stoptime', 'ss_id', 'ss_name', 'ss_latitude', 'ss_longitude', 'es_id', 'es_name', 'es_latitude',
         'es_longitude', 'bikeid', 'usertype', 'birth year', 'gender', 'day_of_week']]

df.head()

tripduration startyear startmonth startdate starttime             stoptime  \
0           695      2013         06        01  00:00:01  2013-06-01 00:11:36   
1           693      2013         06        01  00:00:08  2013-06-01 00:11:41   
2          2059      2013         06        01  00:00:44  2013-06-01 00:35:03   
3           123      2013         06        01  00:01:04  2013-06-01 00:03:07   
4          1521      2013         06        01  00:01:22  2013-06-01 00:26:43   

   ss_id                 ss_name  ss_latitude  ss_longitude  es_id  \
0    444      Broadway & W 24 St    40.742354    -73.989151    434   
1    444      Broadway & W 24 St    40.742354    -73.989151    434   
2    406  Hicks St & Montague St    40.695128    -73.995951    406   
3    475     E 15 St & Irving Pl    40.735243    -73.987586    262   
4   2008   Little West St & 1 Pl    40.705693    -74.016777    310   

                  es_name  es_latitude  es_longitude  bikeid    usertype  \
0         9 Ave & W 18 St    40.743174    -74.003664   19678  Subscriber   
1         9 Ave & W 18 St    40.743174    -74.003664   16649  Subscriber   
2  Hicks St & Montague St    40.695128    -73.995951   19599    Customer   
3         Washington Park    40.691782    -73.973730   16352  Subscriber   
4     State St & Smith St    40.689269    -73.989129   15567  Subscriber   

   birth year  gender  
0        1983       1  
1        1984       1  
2           0       0  
3        1960       1  
4        1983       1

In [15]:
df[['enddate', 'endtime']] = df['stoptime'].str.split(' ', expand=True)

df[['endyear', 'endmonth', 'enddate']] = df['enddate'].str.split('-', expand=True)

df = df[['tripduration', 'startyear', 'startmonth', 'startdate', 'starttime', 'endyear', 'endmonth', 'enddate', 'endtime', 'ss_id', 'ss_name', 'ss_latitude', 'ss_longitude', 'es_id',
         'es_name', 'es_latitude', 'es_longitude', 'bikeid', 'usertype', 'birth year', 'gender', 'day_of_week']]

df.head()

tripduration startyear startmonth startdate starttime endyear endmonth  \
0           695      2013         06        01  00:00:01    2013       06   
1           693      2013         06        01  00:00:08    2013       06   
2          2059      2013         06        01  00:00:44    2013       06   
3           123      2013         06        01  00:01:04    2013       06   
4          1521      2013         06        01  00:01:22    2013       06   

  enddate   endtime  ss_id  ... ss_latitude  ss_longitude  es_id  \
0      01  00:11:36    444  ...   40.742354    -73.989151    434   
1      01  00:11:41    444  ...   40.742354    -73.989151    434   
2      01  00:35:03    406  ...   40.695128    -73.995951    406   
3      01  00:03:07    475  ...   40.735243    -73.987586    262   
4      01  00:26:43   2008  ...   40.705693    -74.016777    310   

                  es_name es_latitude  es_longitude  bikeid    usertype  \
0         9 Ave & W 18 St   40.743174    -74.003664   19678  Subscriber   
1         9 Ave & W 18 St   40.743174    -74.003664   16649  Subscriber   
2  Hicks St & Montague St   40.695128    -73.995951   19599    Customer   
3         Washington Park   40.691782    -73.973730   16352  Subscriber   
4     State St & Smith St   40.689269    -73.989129   15567  Subscriber   

  birth year  gender  
0       1983       1  
1       1984       1  
2          0       0  
3       1960       1  
4       1983       1  

[5 rows x 21 columns]

In [16]:
df.to_csv('clean_data.csv')